lib/EmitVerilog/EmitVerilog.cpp

```diff
@@ -313,16 +316,45 @@ class ModuleEmitter : public VerilogEmitterBase {
   void emitOperation(Operation *op);
 
   void collectNamesEmitDecls(Block &block);
-  void addName(Value value, StringRef name);
-  void addName(Value value, StringAttr nameAttr) {
-    addName(value, nameAttr ? nameAttr.getValue() : "");
+  // Use as key in 'nameTable'. This is an xor field either `Value` can be valid
+  // or `size_t`. If the former, it's any value in the module. If the former, we
+  // need to return the signal name for the result index specified by this
+  // number.
+  using ValueOrResult = std::tuple<Value, size_t>;
```

Comment:
I don't really follow the logic here yet (still reading) but it would be more conventional to use a `PointerIntPair<Value, PointerEmbeddedInt<31>>` or something, since this seems to be a union of two things, not a sum type.

That said, I'm sort of confused why Value isn't enough here.  The only caller of this for result numbers seems to be from the OutputOp processing logic and the argument printing logic.  I would handle this with a different path than the normal 'name' symbol table to avoid making this path more complicated.  Did you consider that approach?

I don't think you have to use this for references to the results of instances -- the `Value` is enough to distinguish between the results being referenced.

Reply:
I did consider separating out the naming for results vs. values. I wanted to reuse the name-conflict-rename logic and this was the first think which occurred to me. I can split that logic out, however, and use it for both the result and value name paths. That would make for simpler code.

Reply:
> I don't think you have to use this for references to the results of instances -- the Value is enough to distinguish between the results being referenced.

Now I remember. I need to associate the SV port name with an identifier for the OutputOp's emit'd `assign` stmt. There are no Values associated with the module results (I don't think), so I use the result index. Is there something else I should use?

Reply:
Yes, I'm suggesting that you change `ModuleEmitter::emitStatement(rtl::OutputOp op)` to not call `getResultName`.  Just push the complexity for this one thing into it, instead of making it core to the whole transformation.

Reply:
Yeah, I agree. I'll still need a way to get the name of the SV port for the assignment. I can't just look it up in the ModuleOp since the output port (result) may be unnamed... Which probably isn't correct in the first place. Ugh. ☹️ I've got two issues to fix, don't I?

Reply:
i'd recommend that you do refactor the conflict renaming stuff out of `ModuleEmitter::addName(Value value, StringRef name)` into a helper function that takes a StringRef for the name and then returns a `StringRef` for the potentially renamified name (pointing into the usedNames set, so it doesn't dangle).  This can be done as a standalone patch (I can do this if it is helpful).

Once you have that, you can then have addName use it, and you can implement a `getOutputName(idx, nameStrIfpresent)` which uses the same helper routine to avoid generating conflicting names.

Reply:
That's exactly what I saying about splitting out the logic. Then I realized I don't want the rename stuff. It's illegal to rename a RTL module port since it breaks instances of it! All I need to do (I think) is reserve the port names at the beginning of the module emit.